programs/test/selftest.c

```diff
@@ -304,6 +305,65 @@ int main( int argc, char *argv[] )
         mbedtls_exit( MBEDTLS_EXIT_FAILURE );
     }
 
+    /*
+     * The C standard allows padding bits in the representation
+     * of standard integer types, but our code does currently not
+     * support them.
+     *
+     * Here we check that the underlying C implementation doesn't
+     * use padding bits, and fail cleanly if it does.
+     *
+     * The check works by casting the maximum value representable
+     * by a given integer type into the unpadded integer type of the
+     * same bit-width and checking that it agrees with the maximum value
+     * of that unpadded type. For example, for a 4-byte int,
+     * MAX_INT should be 0x7fffffff in int32_t. This assumes that
+     * CHAR_BIT == 8, which is checked in check_config.h.
+     */
+
+#define CHECK_PADDING_SIGNED(TYPE, NAME)                                \
+    do                                                                  \
+    {                                                                   \
+        if( ( sizeof( TYPE ) == 2 &&                                    \
+              (int16_t) NAME ## _MAX != 0x7FFF )             ||         \
+            ( sizeof( TYPE ) == 4 &&                                    \
+              (int32_t) NAME ## _MAX != 0x7FFFFFFF )         ||         \
+            ( sizeof( TYPE ) == 8 &&                                    \
+              (int64_t) NAME ## _MAX != 0x7FFFFFFFFFFFFFFF ) )          \
+        {                                                               \
+            mbedtls_printf( "Type '" #TYPE "' has padding bits\n" );    \
+            mbedtls_exit( MBEDTLS_EXIT_FAILURE );                       \
+        }                                                               \
+    } while( 0 )
+
+#define CHECK_PADDING_UNSIGNED(TYPE, NAME)                              \
+    do                                                                  \
+    {                                                                   \
+        if( ( sizeof( TYPE ) == 2 &&                                    \
+              (uint16_t) NAME ## _MAX != 0xFFFF )             ||        \
```

Comment:
Is the check against `..._MAX` sufficient? I have just learned about the bit padding for scalars, and while I understand the check for the `..._MAX` values for both signed and unsigned ints, it is not clear for me, that we can trust that the `..._MIN` values won't be affected by padding.

Reply:
@k-stachowiak If an `s := n + k + 1`-bit wide signed integer type has `k` padding bits, `n` value bits, and `1` sign bits, the maximum expressible value is `2^n - 1`. This equals `2^{s - 1} - 1` if and only if `k=0`, i.e. if and only if there are no padding bits. This is what we check here. The consideration of the maximum value is just one of potentially many means to check the presence of padding bits - we could have considered the minimum value, too, but for that we'd need to know or distinguish between the encoding of negative integers (1's complement vs. 2's complement) which the standard doesn't enforce: We'd have `k=0` if and only if the minimum value is either `-2^{s - 1}` (for 2's complement) or `-(2^{s-1} - 1)` (for 1's complement), and because that check involves two conditions, checking for the maximum value is easier. Also, note that by turning this around, the `XXX_MAX` check above determines the `XXX_MIN` value up to `+-1`. 

Does that help?

Reply:
As far as I understand, all that is based on some assumptions regarding the representation, which, as you stated, is implementation defined. Therefore, if I'm not mistaken, the proposed check is not universal (as it depends on the aforementioned assumptions). Since I don't think there may be illegal (due to the padding), and reside between `x_NIM` and `x_MAX` (i.e. "holes" in the value range), if we also test the difference `x_MAX - x_MIN`, we could detect the presence of the padding bits universally.

Regarding your proposed solution for the signed types, isn't it possible (assuming implementation defined implementation, i.e. basically anything goes), that the `x_MAX` value will pass your test, but the range will be shorter on the `x_MIN` side?

Reply:
@k-stachowiak 

> As far as I understand, all that is based on some assumptions regarding the representation, which, as you stated, is implementation defined.

We rely on:
* The fact that every bit is a value bit, a sign bit, or a padding bit, and that signed/unsigned types have exactly 1/0 sign bits. As @gilles-peskine-arm said, that's part of C99.
* The fact that `intxxx_t` have no padding, which is also mandated by the standard.

Reply:
An implementation might not follow the standard perfectly. But we can't detect all possible departures from the standard and I think the checks here are good enough,